scapy/contrib/automotive/obd/scanner.py

```diff
@@ -0,0 +1,158 @@
+#! /usr/bin/env python
+# flake8: noqa: F405
+
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) Andreas Korb <andreas.korb@e-mundo.de>
+# Copyright (C) Friedrich Feigel <friedrich.feigel@e-mundo.de>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = OnBoardDiagnosticScanner
+# scapy.contrib.status = loads
+
+from scapy.compat import chb
+from scapy.contrib.automotive.obd.obd import *
+
+
+def _supported_id_numbers(socket, timeout, service_class, id_name, verbose):
+    # 'socket' is the ISOTPSocket, over which the OBD-Services communicate.
+    # the id 0x7df acts as a broadcast address for all obd-supporting ECUs.
+    # 'timeout' only required for the OBD Simulator, since it might tell it
+    # supports a PID, while it actually doesn't and won't respond to this PID.
+    # If this happens with a real ECU, it is an implementation error.
+    # 'service class' specifies, which OBD-Service should be queried.
+    # 'id_name' describes the car domain (e.g.: mid = IDs in Motor Domain).
+    # 'verbose' specifies, whether the sr1()-method gives feedback or not.
+
+    supported_id_numbers = set()
+    supported_prop = 'supported_' + id_name + 's'
+
+    # ID 0x00 requests the first range of supported IDs in OBD
+    supported_ids_req = OBD()/service_class(b'\x00')
+
+    while supported_ids_req is not None:
+        resp = socket.sr1(supported_ids_req, timeout=timeout, verbose=verbose)
+
+        # If None, the device did not respond.
+        # Usually only occurs, if device is off.
+        if resp is None:
+            break
+
+        supported_ids_req = None
+
+        all_supported_in_range = getattr(resp.data_records[0], supported_prop)
+
+        for supported in all_supported_in_range:
+            id_number = int(supported[-2:], 16)
+            supported_id_numbers.add(id_number)
+
+            # check whether supported id is for next id range
+            if id_number % 0x20 == 0:
+                supported_ids_req = OBD()/service_class(chb(id_number))
+
+    return supported_id_numbers
+
+
+def _scan_id_service(socket, timeout, service_class, id_numbers, verbose):
+    data = dict()
+
+    for id_number in id_numbers:
+        id_byte = chb(id_number)
+        # assemble request packet
+        pkt = OBD()/service_class(id_byte)
+        resp = socket.sr1(pkt, timeout=timeout, verbose=verbose)
+
+        if resp is not None:
+            data[id_number] = bytes(resp)
+    return data
+
+
+def _scan_dtc_service(socket, timeout, service_class, verbose):
```

Comment:
On thing to consider for your scan and supported functions. The arguments `socket`, `timeout` and `verbose` are just passed through to be used on the call `resp = socket.sr1(pkt, timeout=timeout, verbose=verbose)`.

What about this:
```
def _scan_dtc_service(_sr1, service_class):
    req = OBD()/service_class()
    resp = _sr1(req)
    if resp is not None:
        return bytes(resp)

...
dtc[3] = _scan_dtc_service(lambda pkt: socket.sr1(pkt, timeout=timeout, verbose=verbose), OBD_S03)

# OR
sr1_func = lambda pkt: socket.sr1(pkt, timeout=timeout, verbose=verbose)
dtc[3] = _scan_dtc_service(sr1_func, OBD_S03)
...
supported_ids_s08 = _supported_id_numbers(sr1_func, OBD_S08, 'tid')

```

Just consider it to remove function arguments... If you don't like this.. ignore it.

Reply:
Yes, good point.
In my opinion my current approach seems kind of redundant, while your proposal, though being more elegant, seems to me a bit uneasy to the eye on first sight.